single_cell_analysis/scripts/util/util.R

```diff
@@ -0,0 +1,163 @@
+  } 
+  
+  # convert celltype into numbers
+  celltype_num <- as.numeric(celltype)
+  sample_id <- seq(1:nrow(feature))
+  
+  # iterative k-means
+  # set number of iteration to be 50
+  iter <- 50
+  
+  nmi_score_all <- c()
+  ari_score_all <- c()
+  
+  for(i in 1:iter){
+    # set k equal to the number of celltypes in the dataset
+    k <- length(unique(celltype))
+    # perform k means clustering
+    km <- kmeans(feature, k)
+    # get predicted cluster
+    feature$cluster <- km$cluster
+    
+    # true clusters
+    orignal_data <- data.frame(sample_id, celltype_num)
+    
+    # predicted clusters
+    cl_data <- data.frame(sample_id, feature$cluster)
+    # cacluate NMI and ARI score
+    nmi_score <- NMI::NMI(orignal_data, cl_data)$value
+    ari_score <- mclust::adjustedRandIndex(feature$cluster, celltype_num)
+    
+    nmi_score_all <- c(nmi_score_all, nmi_score)
+    ari_score_all <- c(ari_score_all, ari_score)
+  }
+  
+  return(data.frame(nmi = round(mean(nmi_score_all), 2), ari = round(mean(ari_score_all), 2), 
+                    nmi_var = round(sd(nmi_score_all), 2), ari_var = round(sd(ari_score_all), 2)))
+}
+
+dim_reduc <- function(data, option = "SIMLR", seed = 12345){
+  # Perform SIMLR or PCA dimension reduction 
+  # Args:
+  #   data: matrix used to do dimension reduction
+  #   option: SILMR/PCA 
+  # Returns:
+  #   features projected into two dimensional spaces
+  #
+  set.seed(seed)
+  
+  if(option == "SIMLR"){
+    simlr_result <- SIMLR::SIMLR(X = data, c = 20, cores.ratio = 0)
+    return(simlr_result$ydata)
+  } 
+  
+  else if(option == "PCA"){
+    gene_exp_pca <- as.data.frame(t(data))
+    gene.pca <- prcomp(gene_exp_pca, center = TRUE, scale. = FALSE) 
+    pca_result <- data.frame(gene.pca$x[, 1], gene.pca$x[, 2])
+    return(pca_result)
+  }
+}
+
+ave_sil <- function(feature, celltype){
+  # This function computes the average silhouette score for projected features 
+  # Args:
+  #  feature: a data frame contains projected features (generated by an algorithm, e.g. PCA/ZIFA/VAE),
+  #           and their correpodent cluster (columns), rows are samples.
+  #  celltype: a vector contains cell type information for each sample (cell) 
+  # Returns:
+  #  average silhouette score
+  
+  # convert celltype label into numeric variable
+  cluster <- as.numeric(celltype)
+  
+  # remove the first column that contain sample ID
+  if("id" %in% names(feature)){
+    feature <- feature[, -1]
+  } 
+  
+  # computer silhouette value for each point in a specific cluster
+  ss <- cluster::silhouette(cluster, dist(feature))
+  
+  # average silhouette value
+  ss.mean <- mean(ss[, 3])
+  return(round(ss.mean, 2))
+}
+
+coranking_metric <- function(count.matrix, dm.matrix){
+  # Compute coranking metric to evaluate the performance of dimension reduction
+  # for detail definition of coranking matrix, see 
+  # Lee, J.A., Lee, J.A., Verleysen, M., 2009. Quality assessment of dimensionality reduction:
```

Comment:
This is interesting. Can you add a bit of intuition of how to interpret this coranking metric?